File: SpareChange/src/main/java/com/gocalsd/sparechange/listener/BillingEventListener.java
package com.gocalsd.sparechange.listener;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.android.billingclient.api.BillingResult;
import com.android.billingclient.api.ProductDetails;
import com.android.billingclient.api.Purchase;

import java.util.List;
import java.util.Map;

public interface BillingEventListener {

    /** BillingClient is connected and ready. */
    void onBillingClientReady();

    /** Billing is not available or an operation failed. */
    void onBillingClientUnavailable(@NonNull BillingResult billingResult);

    /** Raw purchase events (new, pending, etc). */
    void onPurchasesUpdated(
            @NonNull BillingResult billingResult,
            @Nullable List<Purchase> purchases
    );

    /** One-time (non-consumable INAPP) license ownership changed. */
    void onOneTimeProductOwnershipChanged(
            @NonNull String productId,
            boolean isOwned
    );

    /** Subscription ownership changed (active or not). */
    void onSubscriptionOwnershipChanged(
            @NonNull String productId,
            boolean isActive
    );

    /**
     * Fired when ProductDetails have been loaded/refreshed
     * and cached by BillingManager.
     */
    void onProductDetailsLoaded(
            @NonNull Map<String, ProductDetails> productDetailsMap
    );

    /**
     * Called when the library detects that the price of a product has changed
     * compared to the last known value stored on this device.
     * Prices are in micros (1,000,000 micros = 1 unit of currency).
     */
    void onProductPriceChanged(
            @NonNull String productId,
            long oldPriceMicros,
            @NonNull String oldCurrency,
            long newPriceMicros,
            @NonNull String newCurrency
    );
}

File: SpareChange/src/main/java/com/gocalsd/sparechange/listener/BillingEventListener.kt
package com.gocalsd.sparechange.listener

import com.android.billingclient.api.BillingResult
import com.android.billingclient.api.ProductDetails
import com.android.billingclient.api.Purchase

interface BillingEventListener {

    /** BillingClient is connected and ready. */
    fun onBillingClientReady()

    /** Billing is not available or an operation failed. */
    fun onBillingClientUnavailable(billingResult: BillingResult)

    /** Raw purchase events (new, pending, etc). */
    fun onPurchasesUpdated(
        billingResult: BillingResult,
        purchases: List<Purchase>?
    )

    /** One-time (non-consumable INAPP) license ownership changed. */
    fun onOneTimeProductOwnershipChanged(
        productId: String,
        isOwned: Boolean
    )

    /** Subscription ownership changed (active or not). */
    fun onSubscriptionOwnershipChanged(
        productId: String,
        isActive: Boolean
    )

    /**
     * Fired when ProductDetails have been loaded/refreshed
     * and cached by BillingManager.
     */
    fun onProductDetailsLoaded(
        productDetailsMap: Map<String, ProductDetails>
    )

    /**
     * Called when the library detects that the price of a product has changed
     * compared to the last known value stored on this device.
     * Prices are in micros (1,000,000 micros = 1 unit of currency).
     */
    fun onProductPriceChanged(
        productId: String,
        oldPriceMicros: Long,
        oldCurrency: String,
        newPriceMicros: Long,
        newCurrency: String
    )

    /**
     * Called when a purchase has been successfully consumed.
     */
    fun onPurchaseConsumed(productId: String, billingResult: BillingResult)

    /**
     * Called when a purchase has been successfully acknowledged.
     */
    fun onPurchaseAcknowledged(productId: String, billingResult: BillingResult)
}
